File: packages/gbnf/javascript/src/grammar-graph/parse-state.ts
import type { Graph, } from "./graph.js";
import type { Pointers, ResolvedRule, } from "./types.js";


export class ParseState extends Function {
  protected _graph: Graph;
  protected _pointers: Pointers;

  constructor(graph: Graph, pointers: Pointers) {
    super();
    this._graph = graph;
    this._pointers = pointers;
    return new Proxy(this, {
      apply: (target, _, args: [string]) => {
        const _call = target._call.bind(target);
        return _call(...args);
      },
    });
  }

  _call(input: string) {
    return this.add(input);
  }

  __call__(input: string): ParseState {
    return this.add(input);
  }

  *[Symbol.iterator](): IterableIterator<ResolvedRule> {
    yield* this.rules();
  }

  *rules(): IterableIterator<ResolvedRule> {
    const rules = new Set<string>();
    for (const { rule, } of this._pointers) {
      const key = JSON.stringify(rule);
      if (!rules.has(key)) {
        rules.add(key);
        yield rule;
      }
    }
  }

  add(input: string): ParseState {
    const pointers = this._graph.add(input, this._pointers);
    return new ParseState(this._graph, pointers);
  }

  get size() {
    return Array.from(this.rules()).length;
  }

  get grammar() {
    return this._graph.grammar;
  }
}

File: packages/gbnf/python/gbnf/grammar_graph/parse_state.py
from __future__ import annotations

from collections.abc import Iterable
from typing import TYPE_CHECKING

from .grammar_graph_types import Rule, RuleType

if TYPE_CHECKING:
    from .grammar_graph_types import Pointers, ResolvedRule
    from .graph import Graph

import json


class CustomJSONEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, RuleType):
            return o.value
        if isinstance(o, Rule):
            return o
        return super().default(o)


class ParseState:
    __graph__: Graph
    __pointers__: Pointers

    def __init__(self, graph: Graph, pointers: Pointers):
        self.__graph__ = graph
        self.__pointers__ = pointers

    def __call__(self, text: str) -> ParseState:
        return self.add(text)

    def __iter__(self) -> Iterable[ResolvedRule]:
        yield from self.rules()

    def rules(self) -> Iterable[ResolvedRule]:
        rules = set()
        for pointer in self.__pointers__:
            rule = pointer.rule
            key = json.dumps(rule.__dict__, cls=CustomJSONEncoder)
            if key not in rules:
                rules.add(key)
                yield rule

    def add(self, text: str) -> ParseState:
        pointers = self.__graph__.add(text, self.__pointers__)
        return ParseState(self.__graph__, pointers)

    def __add__(self, text: str) -> ParseState:
        return self.add(text)

    @property
    def size(self) -> int:
        return len(list(self.rules()))

    @property
    def grammar(self) -> str:
        return self.__graph__.grammar
